README.md
<h1>buneary
<a href="https://circleci.com/gh/dominikbraun/buneary"><img src="https://circleci.com/gh/dominikbraun/buneary.svg?style=shield"></a>
<a href="https://www.codefactor.io/repository/github/dominikbraun/buneary"><img src="https://www.codefactor.io/repository/github/dominikbraun/buneary/badge" /></a>
<a href="https://github.com/dominikbraun/buneary/releases"><img src="https://img.shields.io/github/v/release/dominikbraun/buneary?sort=semver"></a>
<a href="LICENSE"><img src="https://img.shields.io/badge/license-Apache--2.0-brightgreen"></a>
</h1>

`buneary`, pronounced _bun-ear-y_, is an easy-to-use RabbitMQ command line client for managing exchanges, managing
queues, publishing messages to exchanges and reading messages from queues.

<p>
<br>
<img src="logo.png" alt="buneary">
<br>
<br>
</p>

---

## Contents

* [Example](#example)
* [Installation](#installation)
    * [macOS/Linux](#macoslinux)
    * [Windows](#windows)
    * [Docker](#docker)
* [Usage](#usage)
    * [Create an exchange](#create-an-exchange)
    * [Create a queue](#create-a-queue)
    * [Create a binding](#create-a-binding)
    * [Get all exchanges](#get-all-exchanges)
    * [Get an exchange](#get-an-exchange)
    * [Get all queues](#get-all-queues)
    * [Get a queue](#get-a-queue)
    * [Get all bindings](#get-all-bindings)
    * [Get a binding](#get-a-binding)
    * [Get messages in a queue](#get-messages-in-a-queue)
    * [Publish a message](#publish-a-message)
    * [Delete an exchange](#delete-an-exchange)
    * [Delete a queue](#delete-a-queue)
* [Credits](#credits)

## Example

In the following example, a message `Hello!` is published and sent to an exchange called `my-exchange`. The RabbitMQ
server is running on the local machine, and we'll use a routing key called `my-routing-key` for the message.

```
$ buneary publish localhost my-exchange my-routing-key "Hello!"
```

Since the RabbitMQ server listens to the default port, the port can be omitted here. The above command will prompt you
to type in the username and password, but you could do this using command options as well.

## Installation

### macOS/Linux

Download the [latest release](https://github.com/dominikbraun/buneary/releases) for your platform. Extract the
downloaded binary into a directory like `/usr/local/bin`. Make sure the directory is in `PATH`.

### Windows

Download the [latest release](https://github.com/dominikbraun/buneary/releases), create a directory like
`C:\Program Files\buneary` and extract the executable into that directory.
[Add the directory to `Path`](https://www.computerhope.com/issues/ch000549.htm).

### Docker

Just append the actual `buneary` command you want to run after the image name.

Because `buneary` needs to dial the RabbitMQ server, the Docker container needs to be in the same network as the
RabbitMQ server. For example, if the server is running on your local machine, you could run a command as follows:

```
$ docker container run --network=host dominikbraun/buneary version
```

## Usage

### Create an exchange

**Syntax:**

```
$ buneary create exchange <ADDRESS> <NAME> <TYPE> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The desired name of the new exchange.|
|`TYPE`|The exchange type. Has to be one of `direct`, `headers`, `fanout` and `topic`.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|
|`--auto-delete`||Automatically delete the exchange once there are no bindings left.|
|`--durable`||Make the exchange persistent, surviving server restarts.|
|`--internal`||Make the exchange internal.|

**Example:**

Create a direct exchange called `my-exchange` on a RabbitMQ server running on the local machine.

```
$ buneary create exchange localhost my-exchange direct
```

### Create a queue

**Syntax:**

```
$ buneary create queue <ADDRESS> <NAME> <TYPE> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The desired name of the new queue.|
|`TYPE`|The queue type. Has to be one of `classic` and `quorum`.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|
|`--auto-delete`||Automatically delete the queue once there are no consumers left.|
|`--durable`||Make the queue persistent, surviving server restarts.|

**Example:**

Create a classic queue called `my-queue` on a RabbitMQ server running on the local machine.

```
$ buneary create queue localhost my-queue classic
```

### Create a binding

**Syntax:**

```
$ buneary create binding <ADDRESS> <NAME> <TARGET> <BINDING KEY> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The desired name of the new binding.|
|`TARGET`|The name of the target queue or exchange. If it is an exchange, use `--to-exchange`.|
|`BINDING KEY`|The binding key.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|
|`--to-exchange`||Denote that the binding target is another exchange.|

**Example:**

Create a binding from `my-exchange` to `my-queue` on a RabbitMQ server running on the local machine.

```
$ buneary create binding localhost my-exchange my-queue my-binding-key
```

### Get all exchanges

**Syntax:**

```
$ buneary get exchanges <ADDRESS> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Get all exchanges from a RabbitMQ server running on the local machine - this particular example also shows the output.

```
$ buneary get exchanges localhost
User: guest
Password:
+--------------------+---------+---------+-------------+----------+
|        NAME        |  TYPE   | DURABLE | AUTO-DELETE | INTERNAL |
+--------------------+---------+---------+-------------+----------+
|                    | direct  | yes     | no          | no       |
| amq.direct         | direct  | yes     | no          | no       |
| amq.fanout         | fanout  | yes     | no          | no       |
| amq.headers        | headers | yes     | no          | no       |
| amq.match          | headers | yes     | no          | no       |
| amq.rabbitmq.trace | topic   | yes     | no          | yes      |
| amq.topic          | topic   | yes     | no          | no       |
+--------------------+---------+---------+-------------+----------+

```

### Get an exchange

**Syntax:**

```
$ buneary get exchange <ADDRESS> <NAME> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The name of the exchange.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Get an exchange called `my-exchange` from a RabbitMQ server running on the local machine.

```
$ buneary get exchange localhost my-exchange
```

### Get all queues

**Syntax:**

```
$ buneary get queues <ADDRESS> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Get all queues from a RabbitMQ server running on the local machine.

```
$ buneary get queues localhost
```

### Get a queue

**Syntax:**

```
$ buneary get queue <ADDRESS> <NAME> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The name of the queue.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Get a queue called `my-queue` from a RabbitMQ server running on the local machine.

```
$ buneary get queue localhost my-exchange
```

### Get all bindings

**Syntax:**

```
$ buneary get bindings <ADDRESS> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Get all bindings from a RabbitMQ server running on the local machine.

```
$ buneary get bindings localhost
```

### Get a binding

**Syntax:**

```
$ buneary get binding <ADDRESS> <EXCHANGE NAME> <TARGET NAME> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`EXCHANGE NAME`|The name of the source exchange.|
|`TARGET NAME`|The name of the target.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Get the binding or bindings between `my-exchange` and `my-queue` from a RabbitMQ server running on the local machine.

```
$ buneary get binding localhost my-exchange my-queue
```

### Get messages in a queue

**Syntax:**

```
$ buneary get messages <ADDRESS> <QUEUE NAME> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ AMQP address. If no port is specified, `5672` is used.|
|`QUEUE NAME`|The name of the queue to read messages from.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|
|`--max`||The maximum amount of messages to read from the queue.|
|`--requeue`||Reading messages will de-queue them. Re-queue the messages after reading them.|
|`--force`|`-f`|Skip the manual confirmation and force reading the messages.|

**Example:**

Read up to 10 messages from the `my-queue` queue on a RabbitMQ server running on the local machine.

```
$ buneary get messages --max 10 localhost my-queue
```

### Publish a message

**Syntax:**

```
$ buneary publish <ADDRESS> <EXCHANGE> <ROUTING KEY> <BODY> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ AMQP address. If no port is specified, `5672` is used.|
|`EXCHANGE`|The name of the target exchange.|
|`ROUTING KEY`|The routing key of the message.|
|`BODY`|The actual message body.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|
|`--headers`||Comma-separated message headers in the form `--headers key1=val1,key2=val2`.|

**Example:**

Publish a message `Hello!` to `my-exchange` on a RabbitMQ server running on the local machine.

```
$ buneary publish localhost my-exchange my-routing-key "Hello!"
```

### Delete an exchange

**Syntax:**

```
$ buneary delete exchange <ADDRESS> <NAME> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The name of the exchange to be deleted.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Delete an exchange called `my-exchange` on a RabbitMQ server running on the local machine.

```
$ buneary delete exchange localhost my-exchange
```

### Delete a queue

**Syntax:**

```
$ buneary delete queue <ADDRESS> <NAME> [flags]
```

**Arguments:**

|Argument|Description|
|-|-|
|`ADDRESS`|The RabbitMQ HTTP API address. If no port is specified, `15672` is used.|
|`NAME`|The name of the queue to be deleted.|

**Flags:**

|Flag|Short|Description|
|-|-|-|
|`--user`|`-u`|The username to connect with. If not specified, you will be asked for it.|
|`--password`|`-p`|The password to authenticate with. If not specified, you will be asked for it.|

**Example:**

Delete a queue called `my-queue` on a RabbitMQ server running on the local machine.

```
$ buneary delete queue localhost my-queue
```

## Credits

* [michaelklishin/rabbit-hole](https://github.com/michaelklishin/rabbit-hole) is used as RabbitMQ client library.
* [streadway/amqp](https://github.com/streadway/amqp) is used as AMQP client library.
* For all third-party packages used, see [go.mod](go.mod).
* The Buneary graphic is made by [dirocha](https://imgbin.com/user/dirocha).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.1] - 2022-02-16

### Fixed
- Fix constant requeuing of messages (#21).

## [0.3.0] - 2021-02-25

### Added
- Add the `buneary get messages` command.

## [0.2.1] - 2021-02-19

### Changed
- Print success messages on successful resource creations.

## [0.2.0] - 2021-02-10

### Added
- Add the `buneary get exchanges` command.
- Add the `buneary get exchange` command.
- Add the `buneary get queues` command.
- Add the `buneary get queue` command.
- Add the `buneary get bindings` command.
- Add the `buneary get binding` command.
- Add the `--headers` option for specifying message headers.

### Changed
- Use the HTTP API port `15672` instead of the AMQP port `5672`.

## [0.1.1] - 2021-02-07

### Changed
- Enable support for `-v` flag, displaying version information.

### Fixed
- Fix help text for `buneary create binding` command.

## [0.1.0] - 2021-02-05

### First `buneary` release.

## [0.0.0]


jobs.py
import argparse
import sys

CHANGELOG_FILE = "CHANGELOG.md"


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("run", type=str, help="the job to run")
    parser.add_argument("--tag", type=str, help="the Git tag to work with")
    args = parser.parse_args()

    if args.run == "check-changelog":
        check_changelog(args.tag)
    elif args.run == "print-changelog":
        print_changelog(args.tag)

    sys.exit(0)


def check_changelog(git_tag):
    """
    Check if a new release tag is mentioned in the changelog.

    For a release tag like `v1.2.3`, the changelog has to contain a
    release section called `[1.2.3]`. If the release isn't mentioned
    in the changelog, exit with an error.
    """
    # Cut off the `v` prefix to get the actual release number.
    search_expr = "[{0}]".format(git_tag[1:])

    with open(CHANGELOG_FILE) as changelog:
        content = changelog.read()
        if search_expr not in content:
            msg = """You're trying to create a new release tag {0}, but that release is not mentioned
in the changelog. Add a section called {1} to {2} and try again.""" \
                .format(git_tag, search_expr, CHANGELOG_FILE)

            sys.exit(msg)


def print_changelog(git_tag):
    """
    Print the changelog for the given release tag by reading the
    changelog file. If the release tag does not exist as a release
    number in the changelog, the output will be empty.
    """
    start = "## [{0}]".format(git_tag[1:])
    # The ## [Unreleased] heading will be ignored.
    unreleased = "## [Unreleased]"
    end = "## ["

    capturing = False
    output = ""

    with open(CHANGELOG_FILE) as changelog:
        lines = changelog.readlines()
        for line in lines:
            # Start capturing if the line contains our release heading.
            if start in line and unreleased not in line:
                capturing = True
                continue
            # Stop capturing if we've reached the end, i.e. the next heading.
            if end in line and capturing:
                break
            if capturing:
                output += line

    print(output)


main()


buneary.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	rabbithole "github.com/michaelklishin/rabbit-hole/v2"
	"github.com/streadway/amqp"
)

const (
	amqpDefaultPort = 5672
	apiDefaultPort  = 15672
)

type (
	// ExchangeType represents the type of an exchange and thus defines its routing
	// behavior. The type cannot be changed after the exchange has been created.
	ExchangeType string

	// QueueType represents the type of a queue.
	QueueType string

	// BindingType represents the type of a binding and determines whether it binds
	// to a queue - which is the default case - or to another exchange.
	BindingType string
)

const (
	// Direct will deliver messages to queues based on their routing key. A direct
	// exchange compares the routing key to all registered binding keys and forwards
	// the message to all queues with matching binding keys.
	Direct ExchangeType = "direct"

	// Headers will deliver messages to queues based on their headers. This exchange
	// type will ignore the actual routing key.
	Headers = "headers"

	// Fanout will deliver messages to all bound queues of an exchange and ignore
	// the routing key, making them suitable for broadcasting scenarios.
	Fanout = "fanout"

	// Topic will deliver messages to queues based on a binding pattern. The exchange
	// will compare the routing key to all queue binding patterns and forward the
	// message to all matching queues.
	Topic = "topic"

	// Classic represents a classic message queue without any particularities.
	Classic QueueType = "classic"

	// Quorum represents a quorum queue.
	Quorum = "quorum"

	// ToQueue represents a binding from an exchange to a queue.
	ToQueue BindingType = "queue"

	// ToExchange represents a binding from an exchange to another exchange.
	ToExchange = "exchange"
)

// Provider prescribes all functions a buneary implementation has to possess.
type Provider interface {

	// CreateExchange creates a new exchange. If an exchange with the provided name
	// already exists, nothing will happen.
	CreateExchange(exchange Exchange) error

	// CreateQueue will create a new queue. If a queue with the provided name
	// already exists, nothing will happen. CreateQueue will return the queue
	// name generated by the server if no name has been provided.
	CreateQueue(queue Queue) (string, error)

	// CreateBinding will create a new binding. If a binding with the provided
	// target already exists, nothing will happen.
	CreateBinding(binding Binding) error

	// GetExchanges returns all exchanges that pass the provided filter function.
	// To get all exchanges, pass a filter function that always returns true.
	GetExchanges(filter func(exchange Exchange) bool) ([]Exchange, error)

	// GetQueues returns all queues that pass the provided filter function. To get
	// all queues, pass a filter function that always returns true.
	GetQueues(filter func(queue Queue) bool) ([]Queue, error)

	// GetBindings returns all bindings that pass the provided filter function. To
	// get all bindings, pass a filter function that always returns true.
	GetBindings(filter func(binding Binding) bool) ([]Binding, error)

	// GetMessages reads max messages from the given queue. The messages will be
	// re-queued if requeue is set to true. Otherwise, they will be removed from
	// the queue and thus won't be read by subscribers.
	//
	// This behavior may not be obvious to the user, especially if they merely
	// want to "take a look" into the queue without altering its state. Therefore,
	// an implementation should require the user opt-in to this behavior.
	GetMessages(queue Queue, max int, requeue bool) ([]Message, error)

	// PublishMessage publishes a message to the given exchange. The exchange
	// has to exist or must be created before the message is published.
	//
	// The actual message routing is defined by the exchange type. If no routing
	// key is given, the message will be sent to the default exchange.
	PublishMessage(message Message) error

	// DeleteExchange deletes the given exchange from the server. Will return
	// an error if the specified exchange name doesn't exist.
	DeleteExchange(exchange Exchange) error

	// DeleteQueue deletes the given queue from the server. Will return an error
	// if the specified queue name doesn't exist.
	DeleteQueue(queue Queue) error
}

// RabbitMQConfig stores RabbitMQ-related configuration values.
type RabbitMQConfig struct {

	// Address specifies the RabbitMQ address in the form `localhost:5672`. The
	// port is not mandatory. If there's no port, 5672 will be used as default.
	Address string

	// User represents the username for setting up a connection.
	User string

	// Password represents the password to authenticate with.
	Password string
}

// URI returns the AMQP URI for a configuration, prefixed with amqp://.
// In case the RabbitMQ address lacks a port, the default port will be used.
func (a *RabbitMQConfig) URI() string {
	tokens := strings.Split(a.Address, ":")
	var port string

	if len(tokens) == 2 {
		port = tokens[1]
	} else {
		port = strconv.Itoa(amqpDefaultPort)
	}

	uri := fmt.Sprintf("amqp://%s:%s@%s:%s", a.User, a.Password, tokens[0], port)

	return uri
}

// apiURI returns the URI for the RabbitMQ HTTP API, prefixed with http://. In case
// the RabbitMQ server address lacks a port, the default port will be used.
func (a *RabbitMQConfig) apiURI() string {
	tokens := strings.Split(a.Address, ":")
	var port string

	if len(tokens) == 2 {
		port = tokens[1]
	} else {
		port = strconv.Itoa(apiDefaultPort)
	}

	uri := fmt.Sprintf("http://%s:%s", tokens[0], port)

	return uri
}

// Exchange represents a RabbitMQ exchange.
type Exchange struct {

	// Name is the name of the exchange. Names starting with `amq.` denote pre-
	// defined exchanges and should be avoided. A valid name is not empty and only
	// contains letters, digits, hyphens, underscores, periods and colons.
	Name string

	// Type is the type of the exchange and determines in which fashion messages are
	// routed by the exchanged. It cannot be changed afterwards.
	Type ExchangeType

	// Durable determines whether the exchange will be persisted, i.e. be available
	// after server restarts. By default, an exchange is not durable.
	Durable bool

	// AutoDelete determines whether the exchange will be deleted automatically once
	// there are no bindings to any queues left. It won't be deleted by default.
	AutoDelete bool

	// Internal determines whether the exchange should be public-facing or not.
	Internal bool

	// NoWait determines whether the client should wait for the server confirming
	// operations related to the passed exchange. For instance, if NoWait is set to
	// false when creating an exchange, the client won't wait for confirmation.
	NoWait bool
}

// Queue represents a message queue.
type Queue struct {

	// Name is the name of the queue. The name might be empty, in which case the
	// RabbitMQ server will generate and return a name for the queue. Queue names
	// follow the same rules as exchange names regarding the valid characters.
	Name string

	// Type is the type of the queue. Most users will only need classic queues, but
	// buneary strives to support quorum queues as well.
	//
	// For more information, see https://www.rabbitmq.com/quorum-queues.html.
	Type QueueType

	// Durable determines whether the queue will be persisted, i.e. be available after
	// server restarts. By default, an queue is not durable.
	Durable bool

	// AutoDelete determines whether the queue will be deleted automatically once
	// there are no consumers to ready from it left. It won't be deleted by default.
	AutoDelete bool
}

// Binding represents an exchange- or queue binding.
type Binding struct {

	// Type is the type of the binding and determines whether the exchange binds to
	// another exchange or to a queue. Depending on the binding type, the server will
	// look for an exchange or queue with the provided target name.
	Type BindingType

	// From is the "source" of a binding going to the target. Even though this is an
	// Exchange instance, only the exchange name is needed for creating a binding.
	//
	// To bind to a durable queue, the source exchange has to be durable as well. This
	// won't be checked on client-side, but an error will be returned by the server if
	// this constraint is not met.
	From Exchange

	// TargetName is the name of the target, which is either an exchange or a queue.
	TargetName string

	// Key is the key of the binding. The key is crucial for message routing from the
	// exchange to the bound queue or to another exchange.
	Key string
}

// Message represents a message to be enqueued.
type Message struct {

	// Target is the target exchange. Even though this is an entire Exchange instance,
	// only the exchange name is required for sending a message.
	Target Exchange

	// Headers represents the message headers, which is a set of arbitrary key-value
	// pairs. Message headers are considered by some exchange types and thus can be
	// relevant for message routing.
	Headers map[string]interface{}

	// RoutingKey is the routing key of the message and largely determines how the
	// message will be routed and which queues will receive the message. See the
	// individual ExchangeType constants for more information on routing behavior.
	RoutingKey string

	// Body represents the message body.
	Body []byte
}

// NewProvider initializes and returns a default Provider instance.
func NewProvider(config *RabbitMQConfig) Provider {
	b := buneary{
		config: config,
	}
	return &b
}

// buneary is an implementation of the Provider interface with sane defaults.
type buneary struct {
	config  *RabbitMQConfig
	channel *amqp.Channel
	client  *rabbithole.Client
}

// setupChannel dials the configured RabbitMQ server, sets up a connection and opens a
// channel from that connection, which should be closed once buneary has finished.
func (b *buneary) setupChannel() error {
	if b.channel != nil {
		if err := b.channel.Close(); err != nil {
			return fmt.Errorf("closing AMQP channel: %w", err)
		}
	}

	conn, err := amqp.Dial(b.config.URI())
	if err != nil {
		return fmt.Errorf("dialling RabbitMQ server: %w", err)
	}

	if b.channel, err = conn.Channel(); err != nil {
		return fmt.Errorf("establishing AMQP channel: %w", err)
	}

	return nil
}

// setupClient establishes a connection to the RabbitMQ HTTP API, initializing the
// rabbit-hole client. It requires all connection data to exist in the configuration.
func (b *buneary) setupClient() error {
	client, err := rabbithole.NewClient(b.config.apiURI(), b.config.User, b.config.Password)
	if err != nil {
		return fmt.Errorf("creating rabbit-hole client: %w", err)
	}
	b.client = client

	return nil
}

// CreateExchange creates the given exchange. See Provider.CreateExchange for details.
func (b *buneary) CreateExchange(exchange Exchange) error {
	if err := b.setupClient(); err != nil {
		return err
	}

	_, err := b.client.DeclareExchange("/", exchange.Name, rabbithole.ExchangeSettings{
		Type:       string(exchange.Type),
		Durable:    exchange.Durable,
		AutoDelete: exchange.AutoDelete,
	})
	if err != nil {
		return fmt.Errorf("declaring exchange: %w", err)
	}

	return nil
}

// CreateQueue creates the given queue. See Provider.CreateQueue for details.
func (b *buneary) CreateQueue(queue Queue) (string, error) {
	if err := b.setupClient(); err != nil {
		return "", err
	}

	// ToDo: Fetch and return the generated queue name from the response.
	_, err := b.client.DeclareQueue("/", queue.Name, rabbithole.QueueSettings{
		Type:       string(queue.Type),
		Durable:    queue.Durable,
		AutoDelete: queue.AutoDelete,
	})
	if err != nil {
		return "", fmt.Errorf("declaring queue: %w", err)
	}

	return "", nil
}

// CreateBinding creates the given binding. See Provider.CreateBinding for details.
func (b *buneary) CreateBinding(binding Binding) error {
	if err := b.setupClient(); err != nil {
		return err
	}

	_, err := b.client.DeclareBinding("/", rabbithole.BindingInfo{
		Source:          binding.From.Name,
		Vhost:           "/",
		Destination:     binding.TargetName,
		DestinationType: string(binding.Type),
		RoutingKey:      binding.Key,
	})
	if err != nil {
		return fmt.Errorf("declaring binding: %w", err)
	}

	return nil
}

// GetExchanges returns exchanges passing the filter. See Provider.GetExchanges for details.
func (b *buneary) GetExchanges(filter func(exchange Exchange) bool) ([]Exchange, error) {
	if err := b.setupClient(); err != nil {
		return nil, err
	}

	exchangeInfos, err := b.client.ListExchanges()
	if err != nil {
		return nil, fmt.Errorf("listing exchanges: %w", err)
	}

	var exchanges []Exchange

	for _, info := range exchangeInfos {
		e := Exchange{
			Name:       info.Name,
			Type:       ExchangeType(info.Type),
			Durable:    info.Durable,
			AutoDelete: info.AutoDelete,
			Internal:   info.Internal,
		}

		if filter(e) {
			exchanges = append(exchanges, e)
		}
	}

	return exchanges, nil
}

// GetQueues returns queues passing the filter. See Provider.GetQueues for details.
func (b *buneary) GetQueues(filter func(queue Queue) bool) ([]Queue, error) {
	if err := b.setupClient(); err != nil {
		return nil, err
	}

	queueInfos, err := b.client.ListQueues()
	if err != nil {
		return nil, fmt.Errorf("listing queues: %w", err)
	}

	var queues []Queue

	for _, info := range queueInfos {
		q := Queue{
			Name:       info.Name,
			Durable:    info.Durable,
			AutoDelete: info.AutoDelete,
		}

		if filter(q) {
			queues = append(queues, q)
		}
	}

	return queues, nil
}

// GetBindings returns bindings passing the filter. See Provider.GetBindings for details.
func (b *buneary) GetBindings(filter func(binding Binding) bool) ([]Binding, error) {
	if err := b.setupClient(); err != nil {
		return nil, err
	}

	bindingInfos, err := b.client.ListBindings()
	if err != nil {
		return nil, fmt.Errorf("listing bindings: %w", err)
	}

	var bindings []Binding

	for _, info := range bindingInfos {
		b := Binding{
			Type:       BindingType(info.DestinationType),
			From:       Exchange{Name: info.Source},
			TargetName: info.Destination,
			Key:        info.RoutingKey,
		}

		if filter(b) {
			bindings = append(bindings, b)
		}
	}

	return bindings, nil
}

// GetMessages reads messages from the given queue. See Provider.GetMessages for details.
//
// ToDo: Maybe move the function-scoped types somewhere else.
func (b *buneary) GetMessages(queue Queue, max int, requeue bool) ([]Message, error) {
	// getMessagesRequestBody represents the HTTP request body for reading messages.
	type getMessagesRequestBody struct {
		Count    int    `json:"count"`
		Encoding string `json:"encoding"`
		Ackmode  string `json:"ackmode"`
	}

	// getMessagesRequestBody represents the HTTP response body returned by the RabbitMQ
	// API endpoint for reading messages from a queue (/api/queues/vhost/name/get).
	type getMessagesResponseBody []struct {
		PayloadBytes int                    `json:"payload_bytes"`
		Redelivered  bool                   `json:"redelivered"`
		Exchange     string                 `json:"exchange"`
		RoutingKey   string                 `json:"routing_key"`
		Headers      map[string]interface{} `json:"headers"`
		Payload      string                 `json:"payload"`
	}

	ackMode := "ack_requeue_false"
	if requeue {
		ackMode = "ack_requeue_true"
	}

	requestBody := getMessagesRequestBody{
		Count:    max,
		Encoding: "auto",
		Ackmode:  ackMode,
	}

	requestBodyJson, err := json.Marshal(requestBody)
	if err != nil {
		return nil, fmt.Errorf("marshalling request body: %w", err)
	}

	uri := fmt.Sprintf("%s/api/queues/%%2F/%s/get", b.config.apiURI(), queue.Name)

	request, err := http.NewRequest("POST", uri, bytes.NewReader(requestBodyJson))
	if err != nil {
		return nil, fmt.Errorf("creating POST request: %w", err)
	}

	request.SetBasicAuth(b.config.User, b.config.Password)

	response, err := (&http.Client{}).Do(request)
	if err != nil {
		return nil, err
	}

	if response.StatusCode != 200 {
		return nil, fmt.Errorf("RabbitMQ server returned non-200 status: %s", response.Status)
	}

	defer func() {
		_ = response.Body.Close()
	}()

	responseBody := getMessagesResponseBody{}

	if err := json.NewDecoder(response.Body).Decode(&responseBody); err != nil {
		return nil, err
	}

	messages := make([]Message, len(responseBody))

	for i, m := range responseBody {
		messages[i] = Message{
			Target:     Exchange{Name: m.Exchange},
			Headers:    m.Headers,
			RoutingKey: m.RoutingKey,
			Body:       []byte(m.Payload),
		}
	}

	return messages, nil
}

// PublishMessage publishes the given message. See Provider.PublishMessage for details.
func (b *buneary) PublishMessage(message Message) error {
	if err := b.setupChannel(); err != nil {
		return err
	}

	defer func() {
		_ = b.Close()
	}()

	if err := b.channel.Publish(messageArgs(message)); err != nil {
		return fmt.Errorf("publishing message: %w", err)
	}

	return nil
}

// DeleteExchange deletes the given exchange. See Provider.DeleteExchange for details.
func (b *buneary) DeleteExchange(exchange Exchange) error {
	if err := b.setupClient(); err != nil {
		return err
	}

	_, err := b.client.DeleteExchange("/", exchange.Name)
	if err != nil {
		return fmt.Errorf("deleting exchange: %w", err)
	}

	return nil
}

// DeleteQueue deletes the given exchange. See Provider.DeleteQueue for details.
func (b *buneary) DeleteQueue(queue Queue) error {
	if err := b.setupClient(); err != nil {
		return err
	}

	_, err := b.client.DeleteQueue("/", queue.Name)
	if err != nil {
		return fmt.Errorf("deleting queue: %w", err)
	}

	return nil
}

// Close closes the AMQP channel to the configured RabbitMQ server. This function
// should be called after running PublishMessage.
func (b *buneary) Close() error {
	if b.channel != nil {
		if err := b.channel.Close(); err != nil {
			return fmt.Errorf("closing AMQP channel: %w", err)
		}
	}

	return nil
}

// messageArgs returns all message fields expected by the AMQP library as single
// values. This avoids large parameter lists when calling library functions.
func messageArgs(message Message) (string, string, bool, bool, amqp.Publishing) {
	return message.Target.Name,
		message.RoutingKey,
		false,
		false,
		amqp.Publishing{
			Headers:   message.Headers,
			Timestamp: time.Now(),
			Body:      message.Body,
		}
}
